src/policy/policy_deny_fs.cc

```diff
@@ -0,0 +1,137 @@
+#include "policy_deny_fs.h"
+#include "base_object-inl.h"
+#include "v8.h"
+
+#include <fcntl.h>
+#include <limits.h>
+#include <stdlib.h>
+#include <filesystem>
+#include <iostream>
+#include <string>
+#include <vector>
+
+using v8::Just;
+using v8::Maybe;
+
+namespace node {
+
+namespace policy {
+
+// deny = 'fs'
+// deny = 'in:/tmp/'
+// deny = 'in:/tmp/,out:./example.js'
+Maybe<bool> PolicyDenyFs::Apply(const std::string& deny) {
+  for (const auto& name : SplitString(deny, ',')) {
+    Permission perm = Permission::kPermissionsRoot;
+    for (std::string& opt : SplitString(name, ':')) {
+      if (perm == Permission::kPermissionsRoot) {
+        if (opt == "fs") {
+          deny_all_in_ = true;
+          deny_all_out_ = true;
+          return Just(true);
+        }
+        if (opt == "in") {
+          perm = Permission::kFileSystemIn;
+          deny_all_in_ = true;
+        } else if (opt == "out") {
+          perm = Permission::kFileSystemOut;
+          deny_all_out_ = true;
+        } else {
+          return Just(false);
+        }
+      } else {
+        RestrictAccess(perm, opt);
+      }
+    }
+  }
+
+  return Just(true);
+}
+
+bool PolicyDenyFs::Deny(Permission perm, std::vector<std::string> params) {
+  if (perm == Permission::kFileSystem) {
+    deny_all_in_ = true;
+    deny_all_out_ = true;
+    return true;
+  }
+
+  bool deny_all = params.size() == 0;
+  if (perm == Permission::kFileSystemIn) {
+    if (deny_all) deny_all_in_ = true;
+    // when deny_all_in is already true policy.deny should be idempotent
+    if (deny_all_in_) return true;
+
+    RestrictAccess(perm, params);
+    return true;
+  }
+
+  if (perm == Permission::kFileSystemOut) {
+    if (deny_all) deny_all_out_ = true;
+    // when deny_all_out is already true policy.deny should be idempotent
+    if (deny_all_out_) return true;
+
+    RestrictAccess(perm, params);
+    return true;
+  }
+
+  return false;
+}
+
+void PolicyDenyFs::RestrictAccess(Permission perm, const std::string& res) {
+  char resolvedPath[PATH_MAX];
+  // check the result
+  realpath(res.c_str(), resolvedPath);
```

Comment:
Apart from error handling, this is also not quite ideal because it does blocking I/O and doesn’t use libuv. Can this be done in JS land?

Reply:
It would make the design a bit odd. I'll need to check in the `process.policy.deny` if the first parameter is `fs*` and it will be a bit hard to maintain.

Also, this method is called in the environment bootstrap, so... I'd say it's better to use libuv to handle this.